### ISSUE TITLE
`test_import.test_concurrency` leaks ref on Windows

### ISSUE THREAD

**Issue by sunmy2019** [python/cpython#106176]
On Python 3.12, `test_import.test_concurrency()` leaks on Windows:

```
vstinner@WIN C:\victor\python\3.12>python -m test test_import -m test_concurrency -R 3:3
(...)
test_import leaked [22, 36, 28] references, sum=86
```

bisected to #94504

Potentially Related: #104702

See https://github.com/python/cpython/issues/102251#issuecomment-1610523911 https://github.com/python/cpython/pull/104796#issuecomment-1560076330

### Linked PRs
* gh-106207
* gh-108497
* gh-108612

**Comment by vstinner** [comment]
This leak can be seen on the "AMD64 Windows11 Refleaks 3.12" buildbot for example: https://buildbot.python.org/all/#/builders/1103/builds/36

I reproduced it locally.

**Comment by vstinner** [comment]
cc @exarkun @brettcannon

**Comment by Eclips4** [comment]
> On Python 3.12, test_import.test_concurrency() leaks on Windows:

Did you mean that this test leaks only on 3.12?
I can still reproduce it on current main branch (3.13).

**Comment by sunmy2019** [comment]
> Did you mean that this test leaks only on 3.12?

No. It also affects 3.12

**Comment by exarkun** [comment]
I'm not familiar with the leak checking system and I'm not likely to have much time to learn about it in the foreseeable future.

**Comment by brettcannon** [comment]
@exarkun it just means some things are not getting cleaned up.

**Comment by exarkun** [comment]
Just trying to signal my likelihood of pitching on here: low.

**Comment by vstinner** [comment]
Oops, I just reported a duplicate of this bug: https://github.com/python/cpython/issues/107086 (closed).

I also found the same commit with git bisect: commit 3325f054e33b318aa56b74472f76a56b8afc0510

```
commit 3325f054e33b318aa56b74472f76a56b8afc0510
Author: Jean-Paul Calderone <exarkun@twistedmatrix.com>
Date:   Fri Jan 20 19:00:39 2023 -0500

    gh-91351: Fix some bugs in importlib handling of re-entrant imports (GH-94504)

    Co-authored-by: Brett Cannon <brett@python.org>
```

**Comment by vstinner** [comment]
A fix was proposed: PR #106207.

**Comment by erlend-aasland** [comment]
@Yhg1s: IMO, this is also a release blocker (in my opinion, any ref leak is a release blocker).

**Comment by Yhg1s** [comment]
(Yes, this looks like a real leak, not an artifact of how the refleak checker reruns things, so it should be a release blocker.)

**Comment by brettcannon** [comment]
We could revert the fix that leads to the leak to unblock the release as I'm not sure if anyone has come up with a good solution to avoid the memory leak and the bugs that get fixed are not fundamental to import (although it would still be nice to have the bugs fixed).

**Comment by brettcannon** [comment]
I created https://github.com/python/cpython/pull/108497 to try out an idea that @Eclips4 had for using weakrefs.
